File: ScanbotSDKDemo/NFC/PassportReaderViewController.m
//
//  PassportReaderViewController.m
//
//  Created by Sebastian Husche on 30.04.20.
//  Copyright © 2020 doo GmbH. All rights reserved.
//

#import "PassportReaderViewController.h"
#import "PassportReaderResultsViewController.h"
@import ScanbotSDK;

@interface PassportReaderViewController () <SBSDKPassportReaderDelegate, SBSDKUIMRZScannerViewControllerDelegate>

@property(nonatomic, strong) SBSDKNFCPassportReader *nfcReader;
@property(nonatomic, copy) NSString *passportNumber;
@property(nonatomic, copy) NSDate *birthDate;
@property(nonatomic, copy) NSDate *expirationDate;
@property(nonatomic, strong) SBSDKUIMRZScannerViewController *mrzController;
@property(nonatomic, strong) NSArray<SBSDKNFCDatagroup*>* currentResults;
@property (strong, nonatomic) IBOutlet UIProgressView *progressView;


@end

@implementation PassportReaderViewController

- (instancetype)initWithCoder:(NSCoder *)coder {
    if (self = [super initWithCoder:coder]) {
        [self commonInit];
    }
    return self;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    self.modalPresentationStyle = UIModalPresentationFullScreen;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    self.mrzController.recognitionEnabled = YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    SBSDKUIMRZScannerConfiguration *configuration = SBSDKUIMRZScannerConfiguration.defaultConfiguration;
    configuration.uiConfiguration.finderAspectRatio = [[SBSDKAspectRatio alloc]initWithWidth:1.0 andHeight:0.18];
    configuration.uiConfiguration.cancelButtonHidden = YES;
    configuration.textConfiguration.finderTextHint = @"Please align the passports machine readable zone with the frame above to scan it.";
    
    self.mrzController = [SBSDKUIMRZScannerViewController createNewWithConfiguration:configuration andDelegate:self];
    
    [self sbsdk_attachViewController:self.mrzController inView:self.view];
    
    [self.view bringSubviewToFront:self.progressView];
}

- (void)startNFCScanning {
    if (@available(iOS 13, *)) {
        
        self.mrzController.recognitionEnabled = NO;
        self.nfcReader = [[SBSDKNFCPassportReader alloc] initWithPassportNumber:self.passportNumber
                                                                      birthDate:self.birthDate
                                                                 expirationDate:self.expirationDate
                                                                 initialMessage:@"Hold your phone over the passport."
                                                                       delegate:self];
        [self.nfcReader setMessage:@"Authenticating with passport..."];
    }
}

- (void)showResultsIfNeeded {
    if (self.currentResults.count == 0) {
        self.mrzController.recognitionEnabled = YES;
        return;
    }
    PassportReaderResultsViewController *controller = [PassportReaderResultsViewController makeWith:self.currentResults];
    [self.navigationController pushViewController:controller animated:YES];
    self.currentResults = nil;
}

- (void)passportReaderDidConnect:(nonnull SBSDKNFCPassportReader *)reader {
    [self.nfcReader setMessage:@"Enumerating available data groups..."];
    [reader enumerateDatagroups:^(NSArray<SBSDKNFCDatagroupType *>* types) {
        [self.nfcReader setMessage:@"Downloading data groups..."];
        [reader downloadDatagroupsOfType:types completion:^(NSArray<SBSDKNFCDatagroup *>* groups) {
            [self.nfcReader setMessage:@"Finished downloading data groups."];
            if (groups.count > 0) {
                self.currentResults = groups;
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self showResultsIfNeeded];
                });
            }
        }];
    }];
}

- (void)passportReader:(SBSDKNFCPassportReader *)reader didStartReadingGroup:(SBSDKNFCDatagroupType *)type {
    [self.nfcReader setMessage:[NSString stringWithFormat:@"Downloading data group %@...", type]];
    self.progressView.progress = 0.0f;
    self.progressView.hidden = NO;
}

- (void)passportReader:(SBSDKNFCPassportReader *)reader didProgressReadingGroup:(float)progress {
    self.progressView.progress = progress;
    self.progressView.hidden = NO;
}

- (void)passportReader:(SBSDKNFCPassportReader *)reader didFinishReadingGroup:(SBSDKNFCDatagroupType *)type {
    [self.nfcReader setMessage:[NSString stringWithFormat:@"Finished downloading data group %@...", type]];
    self.progressView.progress = 1.0f;
    self.progressView.hidden = YES;
}

- (void)passportReaderDidFinishSession:(SBSDKNFCPassportReader *)reader {
    self.progressView.hidden = YES;
    self.mrzController.recognitionEnabled = YES;
}

- (void)passportReaderDidCancelSession:(SBSDKNFCPassportReader *)reader {
    self.progressView.hidden = YES;
    self.mrzController.recognitionEnabled = YES;
}

- (void)mrzDetectionViewController:(nonnull SBSDKUIMRZScannerViewController *)viewController didDetect:(nonnull SBSDKMachineReadableZoneRecognizerResult *)zone {
    
    if (!self.mrzController.recognitionEnabled) {
        return;
    }
    
    if (zone.documentType != SBSDKMachineReadableZoneRecognizerResultDocumentTypePassport
        || !zone.recognitionSuccessful) {
        
        return;
    }
    
    self.passportNumber = zone.documentCodeField.value;
    
    NSDateFormatter *df = [[NSDateFormatter alloc] init];
    df.locale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
    df.dateFormat = @"dd.MM.yy";
    df.timeZone = [NSTimeZone timeZoneForSecondsFromGMT:0];
    
    self.birthDate = [df dateFromString:zone.dateOfBirthField.value];
    self.expirationDate = [df dateFromString:zone.dateOfExpiryField.value];
    [self startNFCScanning];
}

@end

File: ScanbotSDKDemo/NFC/PassportReaderViewController.swift
//
//  PassportReaderViewController.swift
//  ScanbotSDKDemo
//
//  Created by Danil Voitenko on 18.11.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

import UIKit
import ScanbotSDK

@available(iOS 13.0, *)
class PassportReaderViewController: UIViewController {
    
    private var nfcReader: SBSDKNFCPassportReader?
    private var passportNumber: String?
    private var birthDate: Date?
    private var expirationDate: Date?
    private var mrzController: SBSDKUIMRZScannerViewController?
    private var currentResults: [SBSDKNFCDatagroup]?
    
    @IBOutlet private var progressView: UIProgressView?
    
    private func commonInit() {
        modalPresentationStyle = .fullScreen
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let configuration = SBSDKUIMRZScannerConfiguration.default()
        configuration.uiConfiguration.finderAspectRatio = SBSDKAspectRatio(width: 1.0, andHeight: 0.18)
        configuration.uiConfiguration.isCancelButtonHidden = true
        configuration.textConfiguration.finderTextHint =
        "Please align the passports machine readable zone with the frame above to scan it."
        
        mrzController = SBSDKUIMRZScannerViewController.createNew(with: configuration, andDelegate: self)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        mrzController?.isRecognitionEnabled = true
        guard let mrzController = mrzController, let progressView = progressView else { return }

        sbsdk_attach(mrzController, in: view)
        view.bringSubviewToFront(progressView)
    }
    
    private func startNFCScanning() {
        mrzController?.isRecognitionEnabled = false
        guard let passportNumber = passportNumber,
              let birthDate = birthDate,
              let expirationDate = expirationDate else { return }

        nfcReader = SBSDKNFCPassportReader(passportNumber: passportNumber,
                                        birthDate: birthDate,
                                        expirationDate: expirationDate,
                                        initialMessage: "Hold your phone over the passport.",
                                        delegate: self)
        nfcReader?.setMessage("Authenticating with passport...")
    }
    
    private func showResultsIfNeeded() {
        guard let currentResults = currentResults, !currentResults.isEmpty else {
            mrzController?.isRecognitionEnabled = true
            return
        }
        let controller = PassportReaderResultsViewController.make(with: currentResults)
        navigationController?.pushViewController(controller, animated: true)
        self.currentResults = nil
    }
}

@available(iOS 13.0, *)
extension PassportReaderViewController: SBSDKUIMRZScannerViewControllerDelegate {
    func mrzDetectionViewController(_ viewController: SBSDKUIMRZScannerViewController,
                                    didDetect zone: SBSDKMachineReadableZoneRecognizerResult) {
        if mrzController?.isRecognitionEnabled == false {
            return
        }
        if zone.documentType != SBSDKMachineReadableZoneRecognizerResultDocumentType.passport ||
            !zone.recognitionSuccessful {
            return
        }
        passportNumber = zone.documentCodeField?.value
        
        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.dateFormat = "dd.MM.yy"
        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
        
        birthDate = dateFormatter.date(from: zone.dateOfBirthField!.value)
        expirationDate = dateFormatter.date(from: zone.dateOfExpiryField!.value)
        startNFCScanning()
    }
}

@available(iOS 13.0, *)
extension PassportReaderViewController: SBSDKPassportReaderDelegate {
    
    func passportReaderDidConnect(_ reader: SBSDKNFCPassportReader) {
        nfcReader?.setMessage("Enumerating available data groups...")
        reader.enumerateDatagroups { [weak self] types in
            self?.nfcReader?.setMessage("Downloading data groups...")
            reader.downloadDatagroups(ofType: types) { groups in
                self?.nfcReader?.setMessage("Finished downloading data groups.")
                if !groups.isEmpty {
                    self?.currentResults = groups
                    DispatchQueue.main.async { [weak self] in
                        self?.showResultsIfNeeded()
                    }
                }
            }
        }
    }
    
    func passportReader(_ reader: SBSDKNFCPassportReader, didStartReadingGroup type: String) {
        nfcReader?.setMessage("Downloading data group \(type)...")
        progressView?.progress = 0.0
        progressView?.isHidden = false
    }
    
    func passportReader(_ reader: SBSDKNFCPassportReader, didProgressReadingGroup progress: Float) {
        progressView?.progress = progress
        progressView?.isHidden = false
    }
    
    func passportReader(_ reader: SBSDKNFCPassportReader, didFinishReadingGroup type: String) {
        nfcReader?.setMessage("Finished downloading data group \(type)...")
        progressView?.progress = 1.0
        progressView?.isHidden = true
    }
    
    func passportReaderDidFinishSession(_ reader: SBSDKNFCPassportReader) {
        progressView?.isHidden = true
        mrzController?.isRecognitionEnabled = true
    }
    
    func passportReaderDidCancelSession(_ reader: SBSDKNFCPassportReader) {
        progressView?.isHidden = true
        mrzController?.isRecognitionEnabled = true
    }
}
